### PR TITLE
Improve du

### PR DESCRIPTION
**Pull request by ianthomas23**
Fixes most of #1043.

Changes:

1. Calling `fs.du(path, withdirs=True)` with `path` a directory will include the directory in the output as the first line (i.e. in alphabetical order).
2. Added `withdirs` as an explicit `kwarg` to function to make it easier for users to understand.
3. Docstring improvements.
4. Extended `test_api.py::test_du` and added `test_local.py::test_du`.

There was the question of whether the reported size of a directory should include the size of its contained files and directories. Linux command `du` does include the size of its contents, e.g. using something like `du -Ah -B1 some_directory`. However, I am not sure it is a good idea. It is fine if `maxdepth=None` as the directory walk will fully recurse and can add up directory totals as it walks, but what should happen if `maxdepth=1` for example? Command-line `du` gives the fully-recursed directory size, so if we are aiming for the same functionality here then to get that information we would have to fully recurse, even though the user has specified a `maxdepth`. That seems an undesirable thing to do on a potentially large filesystem. So I have  left it as it was and made it clearer in the docstring that directory size does not include its contents.

### CODE REVIEW

File: fsspec/implementations/tests/test_local.py [hunk 1]
```diff
@@ -823,3 +823,36 @@ def test_put_file_to_dir(tmpdir):
...
+    d = fs.du(tmpdir, total=False, withdirs=True)
+    assert len(d) == 4
+    assert d[file] == 4
+    assert d[subfile] == 7
+    assert d[tmpdir] > 0
+    assert d[subdir] > 0
+
+    assert fs.du(tmpdir, maxdepth=2) == 11
+    assert fs.du(tmpdir, maxdepth=1) == 4
+    assert fs.du(tmpdir, maxdepth=0) == 4
```
Comment (ianthomas23) [inline fsspec/implementations/tests/test_local.py:854]: It surprised me initially that `maxdepth=0` gives the same output as `maxdepth=1`. Maybe it does makes sense though, `maxdepth` effectively has a minimum value of 1.

Comment (martindurant) [reply fsspec/implementations/tests/test_local.py:860]: Is maxdeph <=0 a ValueError then?

Comment (ianthomas23) [reply fsspec/implementations/tests/test_local.py:860]: No, currently `maxdepth <= 1` proceeds as though `maxdepth=1`.

Linux command-line `du` is maybe relevant here. A `maxdepth` less than `0` gives an error, but `maxdepth=0` when specifying a directory path gives you just the directory back without any recursion. So the equivalent behaviour in `fsspec` would be for line 854 above
```python
assert fs.du(tmpdir, maxdepth=0) == 0
```

We could classify this as a bug which should raise an error on `maxdepth<0` and handle `maxdepth=0` differently.

Comment (martindurant) [reply fsspec/implementations/tests/test_local.py:860]: OK, not going to worry about the edge case here. Anyone trying this is just being silly :)
